tiledb/sm/config/config.cc

```diff
@@ -924,39 +937,60 @@ const char* Config::get_from_config(
   return *found ? it->second.c_str() : "";
 }
 
-const char* Config::get_from_config_or_env(
+const char* Config::get_from_config_or_fallback(
     const std::string& param, bool* found) const {
-  // First let's see if the User has set the parameter
-  // If it is not a user set parameter it might be a default value if found in
-  // the config
+  // First check if the user has set the parameter
+  // If not, it may be a default value (if found in the config)
   bool user_set_parameter = set_params_.find(param) != set_params_.end();
 
-  // First check config
+  // [1. user-set config parameters]
   bool found_config;
   const char* value_config = get_from_config(param, &found_config);
-  // If its a user set parameter from the config return it
   if (found_config && user_set_parameter) {
     *found = found_config;
     return value_config;
   }
 
-  // Check if default for this parameter should be ignored based on
-  // environment variables
+  // Check if param default should be ignored based on environment variables
   if (ignore_default_via_env(param)) {
     *found = true;
     return "";
   }
 
-  // Check env if not found in config or if it was found in the config but is
-  // a default value
+  // [2. env variables]
   const char* value_env = get_from_env(param, found);
   if (*found)
     return value_env;
 
-  // At this point the value was not found to be user set in the config or an
-  // environmental variable so return any default value from the config or
-  // indicate it was not found
+  // [3. user-set profiles]
+  auto profile = RestProfile::load_if_exists(
```

Comment:
This potentially does filesystem operations for every single config access, even the ones which cannot possibly be contained in this profile.  We should have a static list of what *can* be found in the profile somewhere so that we can avoid that.

Reply:
Maybe the fact that it "does filesystem operations for every single config access" is a problem in itself.

Reply:
This was resolved as follows:
- When `Config::set_profile` is called, the profile is immediately loaded from disk into memory. After that, only the in-memory profile is used for `Config::get_from_config_or_fallback` calls - no further I/Os done.
- When no explicit profile is set and the user calls `Config::get_from_config_or_fallback` **with a `rest.*` parameter**, the default profile is attempted to be loaded. This means an I/O operation will -inevitably- occur under the hood - but only once, regardless of whether a default profile is found. After that, if a default profile is found, all retrievals of `rest.*` config parameters are handled using the in-memory profile.